### PR TITLE
Feat: Internal staff approve/reject Operation request for carbon tax exemption

### PR DESCRIPTION
**Pull request by andrea-williams**
Feat: Implemented Approve/Reject operation functionality with unit tests

[ZH card 97](https://app.zenhub.com/workspaces/team-moose-sprint-board-60d216175a2689000ec774d0/issues/gh/bcgov/cas-registration/97)

Screenshot 2023-11-01 at 11 50 43 AM

Screenshot 2023-11-01 at 11 51 34 AM

### CODE REVIEW

**Comment by shon-button** [comment]
Nit: Status displays as lower case rather than proper case (approved; rejected) in [Operations](http://localhost:3000/dashboard/operations)

Q: Should the user be able to Submit the form?
If no, then the "Submit" should not display.
If yes, then perhaps the Review component should not be displayed with the confirm message?
image

Q: After Accept\Reject action, should the buttons display on subsequent "View Details" 
If yes, should the Accept\Reject button's disable status reflect the Accept\Reject status
Q: Should the the Accept\Reject status display on subsequent "View Details"

File: bc_obps/registration/api.py [hunk 1]
```diff
@@ -166,12 +191,18 @@ def request_access(request, payload: SelectOperatorIn):
...
+        user=user,
+        operator=operator,
+        role=UserOperator.Roles.ADMIN,
+        status=UserOperator.Statuses.DRAFT,
     )
     return 201, {"user_operator_id": user_operator.id}
 
 
-@router.get("/select-operator/user-operator/{int:user_operator_id}", response=UserOperatorOut)
+@router.get(
```
Comment (JoshLarouche) [inline bc_obps/registration/api.py:202]: It looks like there was a rebase error here that changed the structure of the endpoint

File: bc_obps/registration/api.py [hunk 2]
```diff
@@ -217,7 +248,7 @@ def get_user_operator(request, user_operator_id: int):
 
 
 @router.put(
-    "/select-operator/user-operator/{int:user_operator_id}",
+    "/select-operator/request-access/user-operator/{int:user_operator_id}",
```
Comment (JoshLarouche) [inline bc_obps/registration/api.py:251]: Same rebase error here

File: bc_obps/registration/tests/test_endpoints.py [hunk 3]
```diff
@@ -81,70 +88,157 @@ def test_post_new_operation(self, client):
...
-        )
+        activity = baker.make(ReportingActivity)
+        operation = baker.make(Operation)
+        operation.reporting_activities.set([activity.id])
+
         mock_operation = OperationIn(
-            name='New name',
-            type='Single Facility Operation',
+            name="New name",
+            type="Type 1",
```
Comment (JoshLarouche) [inline bc_obps/registration/tests/test_endpoints.py:213]: rebase error here I think

**Comment by andrea-williams** [comment]
Updates: 

- fixed the rebasing errors that @JoshLarouche  caught (thank you!)
- fixed the bug that @shon-button noticed (great catch!) - the Review component won't appear when creating a new Operation - now it only appears when the Operation's status is Pending (although there maaaaayyyy be a need to change this later, for example, if an internal user accidentally rejects an Operation they intended to Approve, there's no way to undo that)
- the lowercase 'accepted' or 'rejected' status comes from our database. I've added a function to `Operations` component so that the first letter of the status appears capitalized in the table

File: bc_obps/registration/api.py (hunk 2 above)
Comment (JoshLarouche) [reply bc_obps/registration/api.py:265]: Looks like this rebase error is still here

File: .github/workflows/test.yaml [hunk 4]
```diff
@@ -233,6 +233,8 @@ jobs:
         uses: ./.github/actions/local-app-run
         with:
           django_secret_key: ${{ env.DJANGO_SECRET_KEY }}
+      - name: Setup tmate session
+        uses: mxschmitt/action-tmate@v3
```
Comment (JoshLarouche) [inline .github/workflows/test.yaml:237]: I think this needs to be removed

File: bc_obps/registration/api/user_operator.py [hunk 5]
```diff
@@ -63,7 +66,7 @@ def get_user_operator(request, user_operator_id: int):
 
 
 @router.put(
-    "/select-operator/user-operator/{int:user_operator_id}",
+    "/select-operator/request-access/user-operator/{int:user_operator_id}",
```
Comment (JoshLarouche) [inline bc_obps/registration/api/user_operator.py:69]: I think there still might be a rebasing issue here it looks like the route was changed

**Comment by dleard** [comment]
Have we implemented any accessibility checks in our e2e tests? I ask because red/green buttons can be troublesome to get the ratios right to accomodate the most common colour blindness

File: bc_obps/bc_obps/settings.py [hunk 6]
```diff
@@ -26,7 +26,7 @@
 # See https://docs.djangoproject.com/en/4.2/howto/deployment/checklist/
 
 # SECURITY WARNING: keep the secret key used in production secret!
-SECRET_KEY = os.environ.get("DJANGO_SECRET_KEY")
+SECRET_KEY = os.environ.get("SECRET_KEY")
```
Comment (dleard) [inline bc_obps/bc_obps/settings.py:29]: did this change again? It was 'SECRET_KEY', and then a PR last week changed it to 'DJANGO_SECRET_KEY` https://github.com/bcgov/cas-registration/pull/234

File: client/app/components/routes/operations/form/Operation.tsx [hunk 7]
```diff
@@ -4,6 +4,13 @@ import OperationsForm, {
 import { operationSchema } from "@/app/utils/jsonSchema/operations";
 import { RJSFSchema } from "@rjsf/utils";
 import { fetchAPI } from "@/app/utils/api";
+import Review from "./Review";
+<<<<<<< HEAD
+import { Status } from "@/app/types";
+=======
+// 📚 runtime mode for dynamic data to allow build w/o api
+export const runtime = "edge";
+>>>>>>> 1c86db3 (Feat: Implemented Approve/Reject operation functionality with unit tests; 2 other unrelated tests failing)
```
Comment (JoshLarouche) [inline client/app/components/routes/operations/form/Operation.tsx:13]: looks like merge conflict lines got added in here

File: bc_obps/registration/api/operation.py [hunk 8]
```diff
@@ -20,6 +24,18 @@ def get_operation(request, operation_id: int):
...
+    operation = get_object_or_404(Operation, id=operation_id)
+    return operation
+
+
+@router.get("/operations/{operation_id}", response=OperationOut)
+def get_operation(request, operation_id: int):
+    operation = get_object_or_404(Operation, id=operation_id)
+    return operation
+
+
```
Comment (dleard) [inline bc_obps/registration/api/operation.py:38]: Is this all duplicate code? Looks like 3 implementations of the same endpoint?

File: client/app/components/routes/operations/form/Operation.tsx [hunk 9]
```diff
@@ -4,6 +4,13 @@ import OperationsForm, {
 import { operationSchema } from "@/app/utils/jsonSchema/operations";
 import { RJSFSchema } from "@rjsf/utils";
 import { fetchAPI } from "@/app/utils/api";
+import Review from "./Review";
+<<<<<<< HEAD
```
Comment (dleard) [inline client/app/components/routes/operations/form/Operation.tsx:8]: leftover rebase syntax

**Comment by andrea-williams** [comment]
> Have we implemented any accessibility checks in our e2e tests? I ask because red/green buttons can be troublesome to get the ratios right to accomodate the most common colour blindness

at the moment, no. We've got a ticket in our backlog (that will probably end up being split into multiple tickets) to address accessibility accomodations. Right now we're just focused on getting functional features in. Polishing the UI has been de-prioritized for the moment, although we can start thinking about it now that we've got Marcel.

**Comment by dleard** [comment]
> > Have we implemented any accessibility checks in our e2e tests? I ask because red/green buttons can be troublesome to get the ratios right to accomodate the most common colour blindness
> 
> at the moment, no. We've got a ticket in our backlog (that will probably end up being split into multiple tickets) to address accessibility accomodations. Right now we're just focused on getting functional features in. Polishing the UI has been de-prioritized for the moment, although we can start thinking about it now that we've got Marcel.

The 'thumbs up' icon and text (approve/deny) do make it clear what the actions are which mitigates the accessibility issue. Just wanted to flag the green/red buttons as something we may want to revisit. For consistency in CIF, we used a theme throughout the app where all buttons were either solid blue (primary) or outline blue (secondary)

File: bc_obps/bc_obps/settings.py (hunk 6 above)
Comment (andrea-williams) [reply bc_obps/bc_obps/settings.py:29]: this is why I need rebasing coaching! No, it didn't change again. Git is just playing monkey business and I'm too incompetent to catch them at it

**Review by JoshLarouche — APPROVED** [review]
All rebasing leftovers seem to be gone good job! One final nitpick, it looks like your success and reject response messages disappear so fast they aren't really showing the user feedback. Having them disappear after a delay or stay on the page until refresh (I think it may have worked like this before but not sure) would be better.

**Pull request merged**